File: backend/internal/domain/wallpaper.go
package domain

import (
	"context"
	"time"
)

type Wallpaper struct {
	ContestId string
	MimeType  string
	UpdatedAt time.Time
	TextColor string
	Data      []byte
}

type WallpaperRepository interface {
	// Sets the wallpaper data bytes for some contest
	SetWallpaperData(ctx context.Context, contestId string, wallpaper []byte, mimeType string) error

	// Sets the color of the text on the wallpaper, errors if color is not a valid hex string
	SetWallpaperTextColor(ctx context.Context, contestId string, color string) error

	// Deletes the wallpaper for some contest
	DeleteWallpaper(ctx context.Context, contestId string) error

	// Gets the wallpaper for some contest
	GetWallpaper(ctx context.Context, contestId string) (*Wallpaper, error)

	// Gets the last time the wallpaper for some contest was updated to avoid loading the entire wallpaper in memory
	GetLastUpdated(ctx context.Context, contestId string) (time.Time, error)
}

File: backend/src/domain/wallpaper.rs
use async_trait::async_trait;
use chrono::{DateTime, Utc};

use crate::error::AppError;

#[derive(Debug, Clone)]
pub struct Wallpaper {
    pub contest_id: String,
    pub mime_type: String,
    pub updated_at: DateTime<Utc>,
    pub text_color: String,
    pub data: Vec<u8>,
}

#[async_trait]
pub trait WallpaperRepository: Send + Sync {
    async fn set_wallpaper_data(
        &self,
        contest_id: &str,
        data: &[u8],
        mime_type: &str,
    ) -> Result<(), AppError>;
    async fn set_wallpaper_text_color(&self, contest_id: &str, color: &str)
    -> Result<(), AppError>;
    async fn delete_wallpaper(&self, contest_id: &str) -> Result<(), AppError>;
    async fn get_wallpaper(&self, contest_id: &str) -> Result<Option<Wallpaper>, AppError>;
    async fn get_last_updated(&self, contest_id: &str) -> Result<Option<DateTime<Utc>>, AppError>;
}
